components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/WebsocketThriftClient.java

```diff
@@ -0,0 +1,234 @@
+
+		} catch (Exception e) {
+			try {
+
+				log.warn("Login failed.. Authenticating again..");
+				sessionId = thriftUtils.reLogin();
+				//we re-initialize the thrift client in case open sockets have been closed due to
+				//key manager restart.
+				reInitializeClient();
+				thriftDTO = keyValClient
+						.validateKeyforHandshake(context, apiVersion, apiKey, sessionId);
+
+			} catch (Exception e1) {
+				throw new APISecurityException(APISecurityConstants.API_AUTH_GENERAL_ERROR,
+				                               e1.getMessage(), e1);
+			}
+		}
+		//do the conversion other side
+		apiKeyValidationInfoDTO = new APIKeyValidationInfoDTO();
+		apiKeyValidationInfoDTO.setApplicationName(thriftDTO.getApplicationName());
+		apiKeyValidationInfoDTO.setAuthorized(thriftDTO.isAuthorized());
+		apiKeyValidationInfoDTO.setEndUserName(thriftDTO.getEndUserName());
+		apiKeyValidationInfoDTO.setEndUserToken(thriftDTO.getEndUserToken());
+		apiKeyValidationInfoDTO.setSubscriber(thriftDTO.getSubscriber());
+		apiKeyValidationInfoDTO.setTier(thriftDTO.getTier());
+		apiKeyValidationInfoDTO.setType(thriftDTO.getType());
+		apiKeyValidationInfoDTO.setValidationStatus(thriftDTO.getValidationStatus());
+		apiKeyValidationInfoDTO.setApplicationId(thriftDTO.getApplicationId());
+		apiKeyValidationInfoDTO.setApplicationTier(thriftDTO.getApplicationTier());
+		apiKeyValidationInfoDTO.setApiName(thriftDTO.getApiName());
+		apiKeyValidationInfoDTO.setApiPublisher(thriftDTO.getApiPublisher());
+		apiKeyValidationInfoDTO.setConsumerKey(thriftDTO.getConsumerKey());
+		//apiKeyValidationInfoDTO.setAuthorizedDomains(thriftDTO.getAuthorizedDomains());
+		apiKeyValidationInfoDTO.setScopes(thriftDTO.getScopes());
+		apiKeyValidationInfoDTO.setIssuedTime(thriftDTO.getIssuedTime());
+		apiKeyValidationInfoDTO.setValidityPeriod(thriftDTO.getValidityPeriod());
+		apiKeyValidationInfoDTO.setApiTier(thriftDTO.getApiTier());
+		apiKeyValidationInfoDTO.setThrottlingDataList(thriftDTO.getThrottlingDataList());
+		apiKeyValidationInfoDTO.setSpikeArrestLimit(thriftDTO.getSpikeArrestLimit());
+		apiKeyValidationInfoDTO.setSpikeArrestUnit(thriftDTO.getSpikeArrestUnit());
+		apiKeyValidationInfoDTO.setSubscriberTenantDomain(thriftDTO.getSubscriberTenantDomain());
+		apiKeyValidationInfoDTO.setStopOnQuotaReach(thriftDTO.isStopOnQuotaReach());
+		apiKeyValidationInfoDTO.setContentAware(thriftDTO.isIsContentAware());
+		return apiKeyValidationInfoDTO;
+	}
+
+	public ArrayList<URITemplate> getAllURITemplates(String context, String apiVersion)
+			throws APISecurityException {
+		ArrayList<URITemplate> templates = new ArrayList<URITemplate>();
+		List<org.wso2.carbon.apimgt.impl.generated.thrift.URITemplate> uriTemplates;
+
+		try {
+			uriTemplates = keyValClient.getAllURITemplates(context, apiVersion, sessionId);
+
+		} catch (Exception e) {
+			try {
+
+				log.warn("Login failed.. Authenticating again..");
```

Comment:
Its hard to avoid it here because this retry logic should be executed once an exception is cought